internal/docs/design.md

```diff
@@ -104,33 +99,113 @@ https://godoc.org/github.com/google/go-cloud/blob#Bucket.NewWriter
-		can be implemented independently: if one method is mutually exclusive with
-		another, it would be better to return a more complicated data type from one
-		method than to have separate methods.
+    an interface method, then the `error` should implement extra methods and the
+    interface should document these assumptions. Just remember that each method
+    can be implemented independently: if one method is mutually exclusive with
+    another, it would be better to return a more complicated data type from one
+    method than to have separate methods.
 
 -   Transient network errors should be handled by an interface's implementation
-		and not bubbled up as a distinguishable error through a generic interface.
-		Retry logic is best handled as low in the stack as possible to avoid
-		[cascading failure][]. APIs should try to surface "permanent" errors (e.g.
-		malformed request, bad permissions) where appropriate so that application
-		logic does not attempt to retry idempotent operations, but the
-		responsibility is largely on the library, not on the application.
+    and not bubbled up as a distinguishable error through a generic interface.
+    Retry logic is best handled as low in the stack as possible to avoid
+    [cascading failure][]. APIs should try to surface "permanent" errors (e.g.
+    malformed request, bad permissions) where appropriate so that application
+    logic does not attempt to retry idempotent operations, but the
+    responsibility is largely on the library, not on the application.
 
 [cascading failure]:
 https://landing.google.com/sre/book/chapters/addressing-cascading-failures.html
+
+## Tests
+
+For portable APIs, the distinction between developer and operator is blurred for
+integration tests, which require developer-specific resources to be created and
+destroyed. We use [Terraform](http://terraform.io) to do so, and record the
+resource info and network interactions so that they can be replayed as fast and
+repeatable unit tests.
+
+### Replay Mode
+
+Tests normally run in replay mode. In this mode, they don't require any
+provisioned resources or network interactions. Replay tests verify that:
+
+-   The same test inputs produce the same requests (e.g., HTTP requests) to the
+    cloud provider. Some parts of the request may be dynamic (e.g., dates in the
+    HTTP request headers), so the replay tests do some scrubbing when verifying
+    that requests match. Some parts of this scrubbing are provider-specific.
+
+-   The replayed provider responses produce the expected results from the
+    portable API library.
+
+### Record Mode
+
+See https://github.com/google/go-cloud/issues/128 for active work on improving
+this.
+
+1.  Provision resources.
+
+    -   For example, the tests for the AWS implementation of Blob requires a
+        bucket to be provisioned.
+    -   TODO: Use Terraform scripts to provision the resources needed for a
```

Comment:
Add the issue number here, "TODO(issue #)"

Reply:
Done.